### PR TITLE
Make HostSeries overrideable for testing

### PR DESCRIPTION
**Pull request by benhoyt**
This is so we can start using `HostSeries()` instead of `MustHostSeries()` (as part of the effort to make the Juju client not panic on startup for unknown series) and we need to be able to override HostSeries in tests now that we're using that in production code.

Also fixes the unit tests, which had been broken for a while on this branch. Some of them had side effects, so add cleanup functions for those (messy, but oh well). And remove "eoan" from supported now that it's EOL.

### CODE REVIEW

File: series/series.go [hunk 1]
```diff
@@ -20,19 +21,27 @@ const (
...
-	// TODO(katco): Remove globals (lp:1633571)
-	// Override for testing.
+	// HostSeries returns the series of the machine the current process is
+	// running on (overrideable var for testing).
+	HostSeries func() (string, error) = hostSeries
+
+	// MustHostSeries calls HostSeries and panics if there is an error.
+	//
+	// DEPRECATED: don't use this panicky function.
 	MustHostSeries = mustHostSeries
```
Comment (wallyworld) [inline series/series.go:31]: The Must version is fine for tests to use

File: series/supportedseries_test.go [hunk 2]
```diff
@@ -22,6 +24,13 @@ func (s *supportedSeriesSuite) SetUpTest(c *gc.C) {
 	s.CleanupSuite.SetUpTest(c)
 	cleanup := series.SetSeriesVersions(make(map[string]string))
 	s.AddCleanup(func(*gc.C) { cleanup() })
+
+	series.RestoreUbuntuSeries()
+	s.AddCleanup(func(*gc.C) { series.RestoreUbuntuSeries() })
```
Comment (wallyworld) [inline series/supportedseries_test.go:29]: is this intended? don't we just want it in the cleanup?

File: series/series.go [hunk 3]
```diff
@@ -53,7 +62,9 @@ func mustHostSeries() string {
 }
 
 // MustOSFromSeries will panic if the series represents an "unknown"
-// operating system
+// operating system.
+//
+// DEPRECATED: don't use this panicky function.
 func MustOSFromSeries(series string) os.OSType {
```
Comment (wallyworld) [inline series/series.go:68]: Ditto, fine to call from tests

File: series/export_test.go [hunk 4]
```diff
@@ -7,8 +7,23 @@ var (
...
+func init() {
+	origSeriesVersions = make(map[string]string)
+	for k, v := range seriesVersions {
+		origSeriesVersions[k] = v
+	}
+	origUbuntuSeries = make(map[string]seriesVersion)
+	for k, v := range ubuntuSeries {
+		origUbuntuSeries[k] = v
+	}
+}
```
Comment (wallyworld) [inline series/export_test.go:25]: This duplicates what's below in SetSeriesVersions()

File: series/export_test.go [hunk 5]
```diff
@@ -22,7 +37,25 @@ func SetSeriesVersions(value map[string]string) func() {
...
 }
 
+// RestoreSeriesVersions restores the value of seriesVersions to a copy of the
+// original, for use in test cleanup.
+func RestoreSeriesVersions() {
+	seriesVersions = make(map[string]string)
+	for k, v := range origSeriesVersions {
+		seriesVersions[k] = v
+	}
+}
```
Comment (wallyworld) [inline series/export_test.go:47]: SetSeriesVersions() cleanup func does this

File: series/supportedseries_test.go [hunk 6]
```diff
@@ -22,6 +24,13 @@ func (s *supportedSeriesSuite) SetUpTest(c *gc.C) {
 	s.CleanupSuite.SetUpTest(c)
 	cleanup := series.SetSeriesVersions(make(map[string]string))
```
Comment (wallyworld) [inline series/supportedseries_test.go:25]: This is pretty much duplicating what the new restore funcs do

File: series/supportedseries_test.go [hunk 7]
```diff
@@ -22,6 +24,13 @@ func (s *supportedSeriesSuite) SetUpTest(c *gc.C) {
...
 	s.AddCleanup(func(*gc.C) { cleanup() })
+
+	series.RestoreUbuntuSeries()
+	s.AddCleanup(func(*gc.C) { series.RestoreUbuntuSeries() })
+
+	s.PatchValue(series.TimeNow, func() time.Time {
+		return time.Date(2020, 7, 1, 0, 0, 0, 0, time.UTC)
```
Comment (wallyworld) [inline series/supportedseries_test.go:32]: Where do we test for this patched value?

**Review by wallyworld — CHANGES_REQUESTED** [review]
I think there's some duplication of cleanup etc that doesn't seem quite right

File: series/supportedseries_test.go (hunk 7 above)
Comment (benhoyt) [reply series/supportedseries_test.go:30]: We aren't comparing for it in the test itself, it's just that the hard-coded distro info has an EOL date in it, and the code is comparing against that. Specifically the EOL for eoan had elapsed. And I didn't want the unit tests to keep breaking as things go EOL -- we're still testing the right code paths.

File: series/supportedseries_test.go (hunk 2 above)
Comment (benhoyt) [reply series/supportedseries_test.go]: Not intended, thanks.

**Review by wallyworld — APPROVED** [review]

**Comment by benhoyt** [comment]
$$merge$$

**Pull request merged**